### PR TITLE
Fix alignment problem for VectorizedArray

### PR DESCRIPTION
**Pull request by gassmoeller**
@anne-glerum reported a segfault error inside the GMG solver when the Newton terms were enabled. 

I think this patch may fix the reason, but was unable to provoke the error locally on my system (Anne is testing right now). Independent of the result of that test, I think this patch fixes a bug, since `std::vector` does not respect the alignment necessary for `VectorizedArray`. @tjhei does this make sense?

This was introduced in #5580, and was very hard to spot, because it only leads to problems on some runs/vectorization options (whenever the memory allocation is not aligned).

Stacktrace:

```
[bcn1332:145338] *** Process received signal ***
[bcn1332:145338] Signal: Segmentation fault (11)
[bcn1332:145338] Signal code:  (128)
[bcn1332:145338] Failing at address: (nil)
[bcn1332:145338] [ 0] /lib64/libpthread.so.0(+0xf630)[0x2aaae759a630]
[bcn1332:145338] [ 1] /home/bbpanneg/software/aspect/build_release__fix_stresses_elasticity_dealii_9.4.0_on_main_14072023_clean/aspect(_ZSt10__fill_n_aIPN6dealii15VectorizedArrayIdLm8EEEmS2_EN9__gnu_cxx11__enable_ifIXntsrSt11__is_scalarIT1_E7__valueET_E6__typeES9_T0_RKS7_+0x35)[0x32700ea]
[bcn1332:145338] [ 2] /home/bbpanneg/software/aspect/build_release__fix_stresses_elasticity_dealii_9.4.0_on_main_14072023_clean/aspect(_ZSt6fill_nIPN6dealii15VectorizedArrayIdLm8EEEmS2_ET_S4_T0_RKT1_+0x36)[0x326f484]
[bcn1332:145338] [ 3] /home/bbpanneg/software/aspect/build_release__fix_stresses_elasticity_dealii_9.4.0_on_main_14072023_clean/aspect(_ZNSt27__uninitialized_default_n_1ILb1EE18__uninit_default_nIPN6dealii15VectorizedArrayIdLm8EEEmEET_S6_T0_+0x3c)[0x326dd75]
[bcn1332:145338] [ 4] /home/bbpanneg/software/aspect/build_release__fix_stresses_elasticity_dealii_9.4.0_on_main_14072023_clean/aspect(_ZSt25__uninitialized_default_nIPN6dealii15VectorizedArrayIdLm8EEEmET_S4_T0_+0x27)[0x3262e5f]
[bcn1332:145338] [ 5] /home/bbpanneg/software/aspect/build_release__fix_stresses_elasticity_dealii_9.4.0_on_main_14072023_clean/aspect(_ZSt27__uninitialized_default_n_aIPN6dealii15VectorizedArrayIdLm8EEEmS2_ET_S4_T0_RSaIT1_E+0x27)[0x3251423]
[bcn1332:145338] [ 6] /home/bbpanneg/software/aspect/build_release__fix_stresses_elasticity_dealii_9.4.0_on_main_14072023_clean/aspect(_ZNSt6vectorIN6dealii15VectorizedArrayIdLm8EEESaIS2_EE17_M_default_appendEm+0x129)[0x324a883]
[bcn1332:145338] [ 7] /home/bbpanneg/software/aspect/build_release__fix_stresses_elasticity_dealii_9.4.0_on_main_14072023_clean/aspect(_ZNSt6vectorIN6dealii15VectorizedArrayIdLm8EEESaIS2_EE6resizeEm+0x49)[0x3241a0f]
[bcn1332:145338] [ 8] /home/bbpanneg/software/aspect/build_release__fix_stresses_elasticity_dealii_9.4.0_on_main_14072023_clean/aspect(_ZNK6aspect25MatrixFreeStokesOperators14StokesOperatorILi2ELi2EdE11local_applyERKN6dealii10MatrixFreeILi2EdNS3_15VectorizedArrayIdLm8EEEEERNS3_13LinearAlgebra11distributed11BlockVectorIdEERKSD_RKSt4pairIjjE+0x28b)[0x3221f01]
[bcn1332:145338] [ 9] /home/bbpanneg/software/aspect/build_release__fix_stresses_elasticity_dealii_9.4.0_on_main_14072023_clean/aspect(_ZN6dealii8internal8MFWorkerINS_10MatrixFreeILi2EdNS_15VectorizedArrayIdLm8EEEEENS_13LinearAlgebra11distributed11BlockVectorIdEES9_N6aspect25MatrixFreeStokesOperators14StokesOperatorILi2ELi2EdEELb1EE13process_rangeERKMSD_KFvRKS5_RS9_RKS9_RKSt4pairIjjEERKSt6vectorIjSaIjEESW_j+0x304)[0x32785ec]
[bcn1332:145338] [10] /home/bbpanneg/software/aspect/build_release__fix_stresses_elasticity_dealii_9.4.0_on_main_14072023_clean/aspect(_ZN6dealii8internal8MFWorkerINS_10MatrixFreeILi2EdNS_15VectorizedArrayIdLm8EEEEENS_13LinearAlgebra11distributed11BlockVectorIdEES9_N6aspect25MatrixFreeStokesOperators14StokesOperatorILi2ELi2EdEELb1EE4cellEj+0x55)[0x3271c87]
[bcn1332:145338] [11] /home/projects/bbp00039/dealii/dealii_v9.5.2/candi_built/deal.II-v9.5.2/lib/libdeal_II.g.so.9.5.2(_ZNK6dealii8internal19MatrixFreeFunctions8TaskInfo4loopERNS0_17MFWorkerInterfaceE+0x133b)[0x2aaac6c295bb]
[bcn1332:145338] [12] /home/bbpanneg/software/aspect/build_release__fix_stresses_elasticity_dealii_9.4.0_on_main_14072023_clean/aspect(_ZNK6dealii10MatrixFreeILi2EdNS_15VectorizedArrayIdLm8EEEE4loopIN6aspect25MatrixFreeStokesOperators14StokesOperatorILi2ELi2EdEENS_13LinearAlgebra11distributed11BlockVectorIdEESC_EEvMT_KFvRKS3_RT0_RKT1_RKSt4pairIjjEESQ_SQ_PKSD_SH_SK_bNS3_17DataAccessOnFacesEST_+0x108)[0x3239c0a]
[bcn1332:145338] [13] /home/bbpanneg/software/aspect/build_release__fix_stresses_elasticity_dealii_9.4.0_on_main_14072023_clean/aspect(_ZNK6aspect25MatrixFreeStokesOperators14StokesOperatorILi2ELi2EdE9apply_addERN6dealii13LinearAlgebra11distributed11BlockVectorIdEERKS7_+0xc1)[0x31a9afb]
[bcn1332:145338] [14] /home/bbpanneg/software/aspect/build_release__fix_stresses_elasticity_dealii_9.4.0_on_main_14072023_clean/aspect(_ZNK6dealii19MatrixFreeOperators4BaseILi2ENS_13LinearAlgebra11distributed11BlockVectorIdEENS_15VectorizedArrayIdLm8EEEE8mult_addERS5_RKS5_b+0x375)[0x2e7abbf]
[bcn1332:145338] [15] /home/bbpanneg/software/aspect/build_release__fix_stresses_elasticity_dealii_9.4.0_on_main_14072023_clean/aspect(_ZNK6dealii19MatrixFreeOperators4BaseILi2ENS_13LinearAlgebra11distributed11BlockVectorIdEENS_15VectorizedArrayIdLm8EEEE9vmult_addERS5_RKS5_+0x2d)[0x2e1f9f3]
[bcn1332:145338] [16] /home/bbpanneg/software/aspect/build_release__fix_stresses_elasticity_dealii_9.4.0_on_main_14072023_clean/aspect(_ZNK6dealii19MatrixFreeOperators4BaseILi2ENS_13LinearAlgebra11distributed11BlockVectorIdEENS_15VectorizedArrayIdLm8EEEE5vmultERS5_RKS5_+0x46)[0x2dd75d0]
[bcn1332:145338] [17] /home/bbpanneg/software/aspect/build_release__fix_stresses_elasticity_dealii_9.4.0_on_main_14072023_clean/aspect(_ZN6dealii11SolverGMRESINS_13LinearAlgebra11distributed11BlockVectorIdEEE5solveIN6aspect25MatrixFreeStokesOperators14StokesOperatorILi2ELi2EdEENS7_8internal27BlockSchurGMGPreconditionerISA_NS8_14ABlockOperatorILi2ELi2EdEENS8_18MassMatrixOperatorILi2ELi1EdEENS_14PreconditionMGILi2ENS2_6VectorIdNS_11MemorySpace4HostEEENS_20MGTransferMatrixFreeILi2EdEEEESO_EEEEvRKT_RS4_RKS4_RKT0_+0x5b1)[0x2ddac4d]
[bcn1332:145338] [18] /home/bbpanneg/software/aspect/build_release__fix_stresses_elasticity_dealii_9.4.0_on_main_14072023_clean/aspect(_ZN6aspect37StokesMatrixFreeHandlerImplementationILi2ELi2EE5solveEv+0x1fd5)[0x2d19325]
[bcn1332:145338] [19] /home/bbpanneg/software/aspect/build_release__fix_stresses_elasticity_dealii_9.4.0_on_main_14072023_clean/aspect(_ZN6aspect9SimulatorILi2EE12solve_stokesEv+0xe7)[0x2d065ed]
[bcn1332:145338] [20] /home/bbpanneg/software/aspect/build_release__fix_stresses_elasticity_dealii_9.4.0_on_main_14072023_clean/aspect(_ZN6aspect9SimulatorILi2EE36do_one_defect_correction_Stokes_stepERNS_25DefectCorrectionResidualsEb+0xa0c)[0x2dcdc36]
[bcn1332:145338] [21] /home/bbpanneg/software/aspect/build_release__fix_stresses_elasticity_dealii_9.4.0_on_main_14072023_clean/aspect(_ZN6aspect9SimulatorILi2EE42solve_iterated_advection_and_newton_stokesEv+0x494)[0x2d1029a]
[bcn1332:145338] [22] /home/bbpanneg/software/aspect/build_release__fix_stresses_elasticity_dealii_9.4.0_on_main_14072023_clean/aspect(_ZN6aspect9SimulatorILi2EE14solve_timestepEv+0x1cf)[0x2a05233]
[bcn1332:145338] [23] /home/bbpanneg/software/aspect/build_release__fix_stresses_elasticity_dealii_9.4.0_on_main_14072023_clean/aspect(_ZN6aspect9SimulatorILi2EE3runEv+0x5a5)[0x2a0245d]
[bcn1332:145338] [24] /home/bbpanneg/software/aspect/build_release__fix_stresses_elasticity_dealii_9.4.0_on_main_14072023_clean/aspect(_Z13run_simulatorILi2EEvRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES7_bbb+0x2c8)[0x4e08a73]
[bcn1332:145338] [25] /home/bbpanneg/software/aspect/build_release__fix_stresses_elasticity_dealii_9.4.0_on_main_14072023_clean/aspect(main+0x56c)[0x4dd0d62]
[bcn1332:145338] [26] /lib64/libc.so.6(__libc_start_main+0xf5)[0x2aaae77c9555]
[bcn1332:145338] [27] /home/bbpanneg/software/aspect/build_release__fix_stresses_elasticity_dealii_9.4.0_on_main_14072023_clean/aspect[0x29f2e79]
[bcn1332:145338] *** End of error message ***
```

### CODE REVIEW

**Review by tjhei — APPROVED** [review]

**Comment by anne-glerum** [comment]
For the record, it fixed the error for me.

File: source/simulator/stokes_matrix_free.cc [hunk 1]
```diff
@@ -410,8 +410,8 @@ namespace aspect
 
         // Store the symmetric gradients of the velocity field and the
         // values of the pressure field
-        std::vector<SymmetricTensor<2,dim,VectorizedArray<number>>> sym_grad_u;
-        std::vector<VectorizedArray<number>> val_p;
+        AlignedVector<SymmetricTensor<2,dim,VectorizedArray<number>>> sym_grad_u;
+        AlignedVector<VectorizedArray<number>> val_p;
```
Comment (bangerth) [inline source/simulator/stokes_matrix_free.cc:414]: This isn't wrong, but `std::vector` should (and as I far as I know from my experience playing with alignments in deal.II lately) honor alignment. What version of deal.II and what compiler was this with?

Comment (tjhei) [reply source/simulator/stokes_matrix_free.cc:414]: From reading https://stackoverflow.com/a/8456491 it sounds like std::vector alignment is not enough for AVX.

Comment (gassmoeller) [reply source/simulator/stokes_matrix_free.cc:414]: I didnt know about https://stackoverflow.com/a/8456491, all I saw was the explanation here: https://github.com/dealii/dealii/blob/64994cceda58b86f3dbf8b1a19eefa89064db41c/examples/step-37/step-37.cc#L203.

**Pull request merged**